core/src/main/java/co/aikar/commands/RegisteredCommand.java

```diff
@@ -226,8 +226,13 @@ void handleException(CommandIssuer sender, List<String> args, Exception e) {
             if (requiresInput && remainingRequired > 0) {
                 remainingRequired--;
             }
+            String parameterPermission = parameter.getPermission();
             if (args.isEmpty() && !(isLast && type == String[].class)) {
                 if (allowOptional && parameter.getDefaultValue() != null) {
+                    if (parameterPermission != null && !parameterPermission.isEmpty() && !sender.hasPermission(parameterPermission)) {
+                        sender.sendMessage(MessageType.ERROR, MessageKeys.PERMISSION_DENIED);
+                        return null;
+                    }
                     args.add(parameter.getDefaultValue());
                 } else if (allowOptional && parameter.isOptional()) {
                     Object value = parameter.isOptionalResolver() ? resolver.getContext(context) : null;
```

Comment:
The permission check should be removed from the ele and moved here, and checked if isOptionalResolver()

basically, before we execute a .getContext call is when to check, but don't check when the value is null without even calling a resolver.

Reply:
The permission check should also support the multiple perms logic that BaseCommand and RegisteredCommand do.

We should have a Set<String> for getPermission() instead of single string.

The getter in CommandExecutionContext should also represent that as Set<String> return type.

Reply:
> The permission check should be removed from the ele and moved here, and checked if isOptionalResolver()
> 
> basically, before we execute a .getContext call is when to check, but don't check when the value is null without even calling a resolver.

I'm not sure to understand what you want to do by moving the code from the else condition here ? You seem to put it in the condition that checks if the list of arguments is empty, but we want to check the permissions precisely if this list is not empty and so that optional arguments are provided ?

Reply:
Permission should only be checked when resolver.getContext() is called.

Reply:
I made the changes, does that sound good to you ?